src/Tempest/Http/src/Routing/Construction/RouteTreeNode.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Http\Routing\Construction;
+
+use Tempest\Http\Route;
+
+/**
+ * @internal
+ */
+final class RouteTreeNode
```

Comment:
During route discovery I build up a tree (per HTTP method) representing all routes discovered. Each node of this tree can have multiple children and if this node is "the end" of a route, it will also have a `$leaf`  set to the `MarkedRoute`. 

For example, we got 3 routes `/`. `/{id}/edit` and `/new` 
The tree will look something like this:
![tree-example](https://github.com/user-attachments/assets/b04c865d-61d9-4e63-ad78-a0f00f34bd11)

Where only the node `/` -> `{id}` doesn't have a `$leaf` and thus will not be a valid end route. While the others are.

The distinction between `staticPaths` and `dynamicPaths` is also rather useful since we can math the staticPaths before the dynamic ones.